### DISCUSSION TITLE
Allow null-forgiving operator for injected services in Blazor Components

### DISCUSSION THREAD

**Discussion by mwasson74** [tom-englert/Nullable.Extended#14] [Q&A] [answered]
When [Requesting a service in a component](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/dependency-injection?view=aspnetcore-7.0#request-a-service-in-a-component), Microsoft's note suggests:

> ℹ️ Note
> Since injected services are expected to be available, don't mark injected services as nullable. Instead, assign a default literal with the null-forgiving operator (default!). For example:
> ```csharp
> [Inject]
> private IExampleService ExampleService { get; set; } = default!;
> ```

Since I have never worked with analyzers before, would you be able to allow the null-forgiving operator in this case? Or do you have any other suggestions for me to allow this in my code in conjunction with your NuGet package?

Thanks!!

**Comment by tom-englert** [comment]
I'm not sure what you want to do here - what exactly is your problem? Did you read the [manual](https://github.com/tom-englert/Nullable.Extended#roslyn-analyzer)?

**Reply by mwasson74** [reply]
Hello, thanks for getting back to me!

I did read the manual. Since using `default!;` on injected services to a .razor/blazor component is a recommended practice, I was wanting to have the analyzer completely ignore this when/if it sees that the `[Inject]` attribute is being used at the same time in a .razor.cs file. I would like this to be ignored because this is a common practice and I have 100+ instances of this in my Blazor project. I want to set your diagnostic code to error so it cannot be used without justification in all other scenarios. Having to justify 100+ instances on a recommended practice seems rather unnecessary.

Thanks again for your time!!

Matt

**Comment by tom-englert** [comment]
OK, now I see what you mean. 
But wouldn't it be better to fix the root cause, and have a suppression analyzer that suppresses `CS8618` for items that have the `[Inject]` attribute, so you don't have to explicitly use null forgiving?

**Reply by mwasson74** [reply, marked as answer]
That does sound like the best option. Since I haven't the slightest idea how to begin doing that, do you have any suggestions and/or documentation that you could share to help me begin?

Thanks again!!

**Reply by tom-englert** [reply]
Best is to start here: 
https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix
Then this is a sample for a diagnostic suppressor
https://github.com/tom-englert/Nullable.Extended/blob/master/src/Nullable.Extended.Analyzer/Nullable.Extended.Analyzer/NullableDiagnosticSuppressor.cs

**Reply by mwasson74** [reply]
Thank you, I will see what I can come up with!! 🙏

**Reply by tom-englert** [reply]
A better example for a suppression analyzer is this, it better matches your use case:
https://github.com/tom-englert/Lazy.Fody

**Reply by mwasson74** [reply]
Oh, perfect, even better!!
